### PR TITLE
ARROW-3783: [R] Fix float vector conversion

### PR DESCRIPTION
**Pull request by javierluraschi**
https://issues.apache.org/jira/browse/ARROW-3783.

### CODE REVIEW

**Comment by wesm** [comment]
Is it possible to test this (if not, why not)? Ideally we should be able to write

```
arr <- array(1:10, type=float32())
result <- arr$as_vector()
```

**Comment by javierluraschi** [comment]
We would need this fix https://github.com/apache/arrow/pull/2956 to make your example work, but wouldn't help me much since I'm using `RecordBatches` not plain `Arrays`, let me cook the tests using Rcpp and I'll include those here.

**Review by romainfrancois — COMMENTED** [review]
`array()` does not have a `type` argument yet, it's all done with inference for now. 

We rely on [`vctrs::vec_c`](https://vctrs.r-lib.org/reference/vec_c.html) to first convert to an unique R type, this uses the `.ptype` argument, which is kind of doing the same thing than `type` but with R concepts. 

I'm working on `Array$cast` which at least will give us a way to cast an array to an array of `float32`, supporting the `type` argument will require some more work

**Comment by javierluraschi** [comment]
Alright, let me close this one then, thanks @romainfrancois.

**Comment by javierluraschi** [comment]
Sorry, closed the wrong PR. We still need this one, this PR does not make use of `array(type = type)`.

Is there anything you want to change in this PR @romainfrancois?

**Comment by javierluraschi** [comment]
Removed tests, @romainfrancois has a better idea to test this.

File: r/src/array.cpp [hunk 1]
```diff
@@ -945,10 +945,10 @@ SEXP ArrayVector__as_vector(int64_t n, const ArrayVector& arrays) {
       return ArrayVector_To_Vector<Converter_Promotion<REALSXP, arrow::UInt32Type>>(
           n, arrays);
     case Type::HALF_FLOAT:
-      return ArrayVector_To_Vector<Converter_Promotion<REALSXP, arrow::UInt32Type>>(
+      return ArrayVector_To_Vector<Converter_Promotion<REALSXP, arrow::FloatType>>(
```
Comment (romainfrancois) [inline r/src/array.cpp]: actually this needs to be `arrow::HalfFloatType` here. Fixed in #2959

**Comment by wesm** [comment]
Needs rebase. Is this obsolete?

**Comment by romainfrancois** [comment]
I believe this is is obsolete indeed.

**Pull request closed without merging**